Draft: Add recursive deletion support for Draft Layers

tetektoza opened the pull request:
As the title says - this patch implements `onDelete()` method in `ViewProviderLayer` to handle deletion of layer contents. Draft Layers use PropertyLinkListHidden instead of `GroupExtension` so they need to have custom deletion logic.

It adds:
-confirmation dialog for non-empty layers,
-handles `group_recursive_deletion` marker to avoid duplicate dialog logic
-recursively deletes nested layers/groups
-propagates deletion event further to child's `onDelete()` methods

Demo:

https://github.com/user-attachments/assets/06f5c246-04f5-4f33-9779-55d0766b35dd

Note: I didn't intend to change Draft Layers to use `GroupExtension` since it looks like it enforces objects to be only in one group at a time. This kinda feels like breaking purpose of layers, so I've added this custom handling.

Resolves: https://github.com/FreeCAD/FreeCAD/issues/26417

Roy-043 commented:
Thanks. Some initial remarks:
1. A layer should not contain groups. So it is strange, and confusing for future developers, that your code handles them.
2. The title of the dialog for recursive group deletion is wrongly in sentence case. You have matched that in your PR. Maybe @maxwxyz can advise what should be done here?
3. Objects in a layer can be referenced by other objects. Deleting them can be destructive. I suggest checking for that. Note that recursive group deletion currently does not do this either.
4. If you delete a layer containing a PD Body the features of the body remain. It is not just groups that have a Group property.
5. Instead of adding "group_recursive_deletion" to the subelements it would be clearer to add an extra argument to the onDelete function.

semhustej commented:
> 1. A layer should not contain groups. So it is strange, and confusing for future developers, that your code handles them.

I am able to move a group into a layer. It is possible with both 1.0.2 and 1.2dev.

![group-in-layer](https://github.com/user-attachments/assets/8d38953b-338b-4a7e-8c59-257db521c6e1)

I am not sure how practical it is, but it is possible so I guess some users would use it.

Roy-043 commented:
Yes, but changes to the properties of layers are not propagated to objects nested in those groups.

tetektoza commented:
@Roy-043 thanks for the review!

1. I removed that code, although I just added that since there's currently possibility to move groups as @semhustej pointed out.
2. Waiting for Max's feedback
3. I'd leave it like it is in this PR, since groups have the same handling currently, we are not losing anything and I didn't plan to add new functionality to groups, just to extend the existing one to layers.
4. Added, though I had to expose `onDelete` to python so VP knows what's going on. Bodies and their features are being deleted properly now.
5. If you would allow, I'd not touch that for now. It requires changing interface everywhere now, I don't recall why we did this change last time, but I recall it was a workaround. I agree bool is a better choice, but I'd rather want to be sure this PR doesn't touch so many places at once.

Roy-043 commented:
The use of "group_recursive_deletion" is too bizarre IMO. Is it even required or functional after the last modifications? Can you give a scenario where it is? And how does using that string for child objects of the layer make sense?

I do not understand how the subelements argument of the onDelete function is supposed to work normally.

What happens if a user deletes the LayerContainer? Will there be a popup dialog for each layer?

Your checks if an object has already been deleted are strange. You can just use `obj.isAttachedToDocument()` I think.

Roy-043 commented on src/Mod/Draft/draftviewproviders/view_layer.py:
```diff
@@ -525,6 +525,97 @@ def select_contents(self):
+        children = list(self.Object.Group)
+
+        for child in children:
+            if not child or not hasattr(child, "Document"):
```
A document object does not behave as a Boolean: `if not child` does not make sense. And even after deletion it has a `Document` property.
